### Find duplicate samples
The bits of interactive data verification below are done on some intermediate meta data, which looks the same as the published meta_data.csv but has some more info in it ("end", "peak" and "type", which is the manually labeled class).

In [25]:
import datetime as dt
import pandas as pd
df = pd.read_csv('C:/Users/Roman Bolzern/Desktop/D4/neu/train.csv', sep=";", parse_dates=["start", "end", "peak"], index_col="id")
timesteps = pd.TimedeltaIndex([dt.timedelta(minutes=d) for d in [0, 7*60, 10*60+30, 11*60+50]])
all_image_times = []
for t in timesteps:
    new_df = df[['start', 'noaa_num']].copy()
    new_df['start'] = new_df['start'] + t
    all_image_times.append(new_df)
all_image_times = pd.concat(all_image_times)
res = all_image_times.groupby(by=['noaa_num', 'start']).size().reset_index(name='counts')
print(f"{len(res[res['counts'] > 1])} of {len(all_image_times)} images have a duplicate")
res[res['counts'] > 1]

26 of 33348 images have a duplicate


noaa_num               start  counts
417       11401 2012-01-22 01:33:43       2
1757      11450 2012-04-03 04:27:01       2
3308      11504 2012-06-19 04:28:01       2
4264      11528 2012-07-28 23:24:01       2
6628      11635 2012-12-21 11:31:00       2
6629      11635 2012-12-21 18:31:00       2
6630      11635 2012-12-21 22:01:00       2
6631      11635 2012-12-21 23:21:00       2
7485      11663 2013-01-30 03:04:01       2
9708      11743 2013-05-11 08:53:01       2
10742     11800 2013-07-19 22:08:01       2
11372     11834 2013-09-02 12:51:01       2
11742     11850 2013-09-22 21:09:01       2
14799     11996 2014-03-09 22:50:01       2
15519     12026 2014-04-01 01:08:01       2
21675     12253 2015-01-02 04:47:01       2
29304     12543 2016-05-13 07:22:00       2
29314     12543 2016-05-13 14:22:00       2
29318     12543 2016-05-13 17:52:00       2
29319     12543 2016-05-13 19:12:00       2
29355     12544 2016-05-13 07:22:00       2
29359     12544 2016-05-13 14:22:00       2
29362     12544 2016-05-13 17:52:00       2
29363     12544 2016-05-13 19:12:00       2
29805     12565 2016-07-13 21:28:01       2
30647     12599 2016-10-14 23:35:01       2

Not too wild, but AR 11635 looks suspicious:

In [26]:
df['peak_after_end'] = df['peak'] - df['end']
df[df['noaa_num'] == 11635].sort_values(by=['start'])

noaa_num               start                 end  \
id                                                                              
11635_2012_12_19_12_00_00_0     11635 2012-12-19 00:00:00 2012-12-19 12:00:00   
11635_2012_12_19_12_00_00_1     11635 2012-12-19 16:52:40 2012-12-20 04:52:40   
11635_2012_12_20_04_54_01_0     11635 2012-12-19 16:54:01 2012-12-20 04:54:01   
11635_2012_12_20_17_08_01_0     11635 2012-12-20 05:08:01 2012-12-20 17:08:01   
11635_2012_12_20_19_34_01_0     11635 2012-12-20 07:34:01 2012-12-20 19:34:01   
11635_2012_12_20_20_41_01_0     11635 2012-12-20 08:41:01 2012-12-20 20:41:01   
11635_2012_12_20_04_54_01_1     11635 2012-12-20 11:53:28 2012-12-20 23:53:28   
11635_2012_12_20_23_55_01_0     11635 2012-12-20 11:55:01 2012-12-20 23:55:01   
11635_2012_12_21_04_05_01_0     11635 2012-12-20 16:05:01 2012-12-21 04:05:01   
11635_2012_12_20_23_55_01_1     11635 2012-12-21 11:31:00 2012-12-21 23:31:00   
11635_2012_12_21_04_05_01_1     11635 2012-12-21 11:31:00 2012-12-21 23:31:00   
11635_2012_12_22_23_38_00_0     11635 2012-12-22 11:38:00 2012-12-22 23:38:00   
11635_2012_12_23_02_41_11_0     11635 2012-12-22 14:41:11 2012-12-23 02:41:11   
11635_2012_12_23_02_41_34_0     11635 2012-12-22 14:41:34 2012-12-23 02:41:34   
11635_2012_12_23_02_41_34_1     11635 2012-12-23 04:11:07 2012-12-23 16:11:07   
11635_2012_12_23_16_14_01_0     11635 2012-12-23 04:14:01 2012-12-23 16:14:01   
11635_2012_12_24_06_43_01_0     11635 2012-12-23 18:43:01 2012-12-24 06:43:01   
11635_2012_12_23_16_14_01_1     11635 2012-12-23 22:26:00 2012-12-24 10:26:00   
11635_2012_12_25_10_52_00_0     11635 2012-12-24 22:52:00 2012-12-25 10:52:00   
11635_2012_12_26_04_56_47_0     11635 2012-12-25 16:56:47 2012-12-26 04:56:47   
11635_2012_12_26_04_57_30_0     11635 2012-12-25 16:57:30 2012-12-26 04:57:30   
11635_2012_12_26_04_57_59_0     11635 2012-12-25 16:57:59 2012-12-26 04:57:59   
11635_2012_12_27_02_27_53_0     11635 2012-12-26 14:27:53 2012-12-27 02:27:53   
11635_2012_12_27_13_25_00_0     11635 2012-12-27 01:25:00 2012-12-27 13:25:00   
11635_2012_12_27_13_25_00_1     11635 2012-12-27 23:10:00 2012-12-28 11:10:00   
11635_2012_12_29_02_43_01_0     11635 2012-12-28 14:43:01 2012-12-29 02:43:01   

                             type                peak     peak_flux  \
id                                                                    
11635_2012_12_19_12_00_00_0  B4.9 2012-12-20 09:46:00  5.764706e-07   
11635_2012_12_19_12_00_00_1  B4.9 2012-12-20 09:46:00  5.764706e-07   
11635_2012_12_20_04_54_01_0  B6.0 2012-12-21 04:54:00  7.058824e-07   
11635_2012_12_20_17_08_01_0  B5.4 2012-12-21 17:08:00  6.352941e-07   
11635_2012_12_20_19_34_01_0  B6.5 2012-12-21 19:34:00  7.647059e-07   
11635_2012_12_20_20_41_01_0  B6.4 2012-12-21 20:41:00  7.529412e-07   
11635_2012_12_20_04_54_01_1  B6.0 2012-12-21 04:54:00  7.058824e-07   
11635_2012_12_20_23_55_01_0  C1.9 2012-12-21 23:55:00  2.235294e-06   
11635_2012_12_21_04_05_01_0  C1.9 2012-12-22 04:05:00  2.235294e-06   
11635_2012_12_20_23_55_01_1  C1.9 2012-12-21 23:55:00  2.235294e-06   
11635_2012_12_21_04_05_01_1  C1.9 2012-12-22 04:05:00  2.235294e-06   
11635_2012_12_22_23_38_00_0  C1.5 2012-12-23 02:30:00  1.764706e-06   
11635_2012_12_23_02_41_11_0  B7.5 2012-12-23 14:29:00  8.823529e-07   
11635_2012_12_23_02_41_34_0  B9.3 2012-12-23 20:44:00  1.094118e-06   
11635_2012_12_23_02_41_34_1  B9.3 2012-12-23 20:44:00  1.094118e-06   
11635_2012_12_23_16_14_01_0  C1.7 2012-12-24 16:14:00  2.000000e-06   
11635_2012_12_24_06_43_01_0  C1.8 2012-12-25 06:43:00  2.117647e-06   
11635_2012_12_23_16_14_01_1  C1.7 2012-12-24 16:14:00  2.000000e-06   
11635_2012_12_25_10_52_00_0  C4.1 2012-12-25 13:04:00  4.823529e-06   
11635_2012_12_26_04_56_47_0  C1.6 2012-12-27 02:24:00  1.882353e-06   
11635_2012_12_26_04_57_30_0  C1.6 2012-12-26 11:25:00  1.882353e-06   
11635_2012_12_26_04_57_59_0  C1.6 2012-12-26 17:14:00  1.882353e-06   
11635_2012_12_27_02_27_53_0  B9.4 2012-12-27 13:20:00  1

Two successive row groups jump into my view...

In [27]:
df[(df['noaa_num'] == 11635) & (df['start'] == pd.Timestamp('2012-12-21 11:31:00'))].sort_values(by=['start'])

noaa_num               start                 end  \
id                                                                              
11635_2012_12_21_04_05_01_1     11635 2012-12-21 11:31:00 2012-12-21 23:31:00   
11635_2012_12_20_23_55_01_1     11635 2012-12-21 11:31:00 2012-12-21 23:31:00   

                             type                peak  peak_flux  \
id                                                                 
11635_2012_12_21_04_05_01_1  C1.9 2012-12-22 04:05:00   0.000002   
11635_2012_12_20_23_55_01_1  C1.9 2012-12-21 23:55:00   0.000002   

                            peak_after_end  
id                                          
11635_2012_12_21_04_05_01_1       04:34:00  
11635_2012_12_20_23_55_01_1       00:24:00

In [28]:
df[(df['noaa_num'] == 11635) & (df['start'].dt.day == 25)].sort_values(by=['start'])

noaa_num               start                 end  \
id                                                                              
11635_2012_12_26_04_56_47_0     11635 2012-12-25 16:56:47 2012-12-26 04:56:47   
11635_2012_12_26_04_57_30_0     11635 2012-12-25 16:57:30 2012-12-26 04:57:30   
11635_2012_12_26_04_57_59_0     11635 2012-12-25 16:57:59 2012-12-26 04:57:59   

                             type                peak  peak_flux  \
id                                                                 
11635_2012_12_26_04_56_47_0  C1.6 2012-12-27 02:24:00   0.000002   
11635_2012_12_26_04_57_30_0  C1.6 2012-12-26 11:25:00   0.000002   
11635_2012_12_26_04_57_59_0  C1.6 2012-12-26 17:14:00   0.000002   

                            peak_after_end  
id                                          
11635_2012_12_26_04_56_47_0       21:27:13  
11635_2012_12_26_04_57_30_0       06:27:30  
11635_2012_12_26_04_57_59_0       12:16:01

Seems like the first sample is for two separate C1.9 flares, but cut the same way. Maybe there are flares before and after this range so that there aren't any other cutting options. Imho not worth of further investigation.
The second group's ranges are overlapping heavily, even though their peaks differ. Same issue as before it seems. One might ask, are there many such overlapping samples in the dataset?

In [49]:
gp = df.sort_values(by=['start']).groupby(['noaa_num'])
for g in gp:
    df.loc[g[1].index, 'delta'] = (g[1]['start']-g[1]['start'].shift())

In [57]:
threshold = 30
overlapping_samples = df[df['delta'] < pd.Timedelta(minutes=threshold)]
print(f"{len(overlapping_samples)} of {len(df)} samples have a sample start that is followed directly by another sample start (same AR, max {threshold} minutes afterwards)")

583 of 8337 samples have a sample start that is followed directly by another sample start (same AR, max 30 minutes afterwards)


Tbh, 583 strongly overlapping samples are a few more than I expected. It's nothing too bad and their values seem correct, it might be worth to investigate into improving those overlaps at some point though.
The 26 (almost) duplicate samples aren't too wild, and by taking care of the 583 overlaps those 26 would be taken care of too.